### DISCUSSION TITLE
(🎁) How to subtype date?

### DISCUSSION THREAD

**Discussion by KotlinIsland** [pydantic/pydantic#8122] [Question]
Pydantic will automatically convert strings into dates:
```py
TypeAdapter(date).validate_python("2023-11-15")  # datetime.date(2023, 11, 15)
```

But for a subtype of date this doesn't work:
```py
class MyDate(date): ...
TypeAdapter(MyDate).validate_python("2023-11-15")  # pydantic.errors.PydanticSchemaGenerationError: Unable to generate pydantic-core schema for <class '__main__.MyDate'>. Set `arbitrary_types_allowed=True` in the model_config to ignore this error or implement `__get_pydantic_core_schema__` on your type to fully support it.
```

Adding a naïve implementation of this method results in a `datetime.date`, not a `MyDate`:
```py
class MyDate(date):
    def __get_pydantic_core_schema__(self):
        return core_schema.date_schema()

TypeAdapter(MyDate).validate_python("2023-11-15")  # datetime.date(2023, 11, 15)
```

I tried an implementation with `Annotated`, but it's not desirable, as now I have two objects that should only be functionality contained in the class.
```py
class _MyDate(date):
    @classmethod
    def from_date(cls, d: date) -> Self:
        return cls(d.year, d.month, d.day)

    def __get_pydantic_core_schema__(self):
        return core_schema.date_schema()

MyDate = Annotated[_MyDate, AfterValidator(_MyDate.from_date)]
TypeAdapter(MyDate).validate_python("2023-11-15")  # _MyDate(2023, 11, 15)
```

tl;dr: I want to subclass `date`, and I want to inherit pydantics date conversion logic.

**Comment by joemar25** [comment]
Tricky, Pydantic’s validation and conversion logic is tied to specific types, and it doesn’t automatically apply to subclasses. One possible solution is to create a custom data type that wraps the date type instead of subclassing it. Try this:

```python
from datetime import date
from pydantic import BaseModel, field_validator
from pydantic_core import CoreSchema
from pydantic import GetJsonSchemaHandler


class MyDateType(date):
    @classmethod
    def __get_pydantic_json_schema__(
        cls, core_schema: CoreSchema, handler: GetJsonSchemaHandler
    ):
        json_schema = super().__get_pydantic_json_schema__(core_schema, handler)
        json_schema.update(type="string", format="date")
        return json_schema

    @classmethod
    def validate(cls, v):
        if not isinstance(v, date):
            raise TypeError("datetime.date required")
        return cls(v.year, v.month, v.day)


class MyDateModel(BaseModel):
    value: MyDateType

    class Config:
        arbitrary_types_allowed = True

    @field_validator("value", mode="before")
    def parse_date(cls, v):
        if isinstance(v, str):
            return MyDateType.fromisoformat(v)
        return v


m = MyDateModel(value="2023-11-15")
print(m.value)
```

This approach allows you to create a subclass of date that can be used with Pydantic’s TypeAdapter and automatically convert date strings to date objects.

Sources:
https://stackoverflow.com/questions/66548586/how-to-change-date-format-in-pydantic
https://stackoverflow.com/questions/76569003/pydantic-custom-data-type-that-subclasses-built-in-list
https://stackoverflow.com/questions/72425843/pydantic-basemodel-field-for-validation-purposes-only
https://docs.pydantic.dev/2.5/errors/usage_errors/#custom-json-schema

**Comment by KotlinIsland** [comment]
```py
class MyDate(date):
    @classmethod
    def __get_pydantic_core_schema__(
        cls, source_type: type[object], handler: GetCoreSchemaHandler
    ) -> CoreSchema:
        return core_schema.union_schema(
            [
                core_schema.no_info_plain_validator_function(MyDate),
                core_schema.no_info_plain_validator_function(MyDate.fromisoformat),
            ]
        )
```

Something like this seems to work.
